[WD-16233] update Jira task statuses

akbarkz opened the pull request:
## Done

 - Add a background task that fetches existing Jira tasks' statuses and updates them in the database
 - Sort Jira tasks on a webpage to show latest ones first

## QA

### QA steps

- Run the app locally using `dotrun` or use the demo link below
- Create a Jira task on any webpage by either clicking on `Request changes` or `Request removal` buttons. You can also click on `Request new page` button from the sidebar, fill in all the required fields, save it, and then click on `Submit for publication..." button. Any of these actions will work, cause they all create a Jira task which is the goal here.
- After creating a Jira task, make sure it appears on a webpage under "Related Jira tickets" section.
- Open the Jira ticket by clicking on its ID and change its status from Jira. 
- Wait for 5 minutes (that's the default delay, you can change it locally in `.env.local` file by setting the of UPDATE_STATUS_DELAY env var to a lower number)
- Refresh the page and check if the Jira task status has been updated in the Content System.
- Create another Jira task on the same webpage.
- Check that the newly created Jira task shows first, i.e. the list is ordered correctly. 

## Fixes

 - Fixes https://warthogs.atlassian.net/browse/WD-16233

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://cs-canonical-com-118.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/12705)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

muhammad-ali-pk commented on .env:
```diff
@@ -6,6 +6,7 @@ GH_TOKEN=token
 REPO_ORG=https://github.com/canonical
 DATABASE_URL=postgresql://postgres:postgres@localhost:5432/content
 TASK_DELAY=30
+STATUS_UPDATE_DELAY=5
```
```suggestion
UPDATE_STATUS_DELAY=5
```

muhammad-ali-pk commented:
@akbarkz 
Opened up an existing page, then submitted request changes. It created a Jira task. I updated the status of Jira task from within Jira. The background task ran after the specified duration, but it didn't change the status of the webpage, even after page refresh.

muhammad-ali-pk commented on webapp/jira.py:
```diff
@@ -254,6 +254,21 @@ def change_issue_status(self, issue_id: str, transition_id: str) -> bool:
+        Args:
+            task (JiraTask): The Jira task instance.
+        """
+        response = self.__request__(
+            method="GET",
+            url=f"{self.url}/rest/api/3/issue/{task.jira_id}?fields=status",
+        )
```
Might be a good idea to extract this logic into it's own method. Can be reused in future to fetch jira tasks, as well as promotes Single Responsibility Principle.

akbarkz replied:
Yes I agree with you, it would be better to have a designated file that has all the methods that call Jira API only. I would skip doing that yet, cause that would require refactoring the whole file. For now we have this `jira.py` helper functions file that deals with Jira.

akbarkz commented:
> @akbarkz Opened up an existing page, then submitted request changes. It created a Jira task. I updated the status of Jira task from within Jira. The background task ran after the specified duration, but it didn't change the status of the Jira task on webpage, even after page reload.
> 
> [EDIT]: Upon further investigation, it looks like the status is updated in the database, but not being reflected on the webpage itself.

Good catch! Apparently there was an error of how users get added to `users` table, causing duplication and consequent errors when fetching data. I fixed that by changing the logic of identifying whether a user exists or not, for that I use their emails rather than HRC ids like before. That fixed the issue partially though. I also had to add invalidation of the cache after updating the statuses.

muhammad-ali-pk commented on webapp/jira.py:
```diff
@@ -255,6 +257,43 @@ def change_issue_status(self, issue_id: str, transition_id: str) -> bool:
+            app (Flask): The Flask application instance.
+            task_locks (dict): The locks for the project trees.
+        """
+        jira_tasks = JiraTask.query.all()
+        if jira_tasks:
+            project_ids = []
+            for task in jira_tasks:
+                response = self.__request__(
+                    method="GET",
+                    url=f"{self.url}/rest/api/3/issue/{task.jira_id}?fields=status",
+                )
+                if task.status != response["fields"]["status"]["name"].upper():
+                    task.status = response["fields"]["status"]["name"].upper()
+                    # get the project id from the webpage that corresponds to
+                    # the Jira task (will be needed to invalidate the cache)
+                    webpage = Webpage.query.filter_by(
+                        id=task.webpage_id
+                    ).first()
+                    if webpage.project_id not in project_ids:
+                        project_ids.append(webpage.project_id)
+            db.session.commit()
```
Wondering if this is the best approach to updating Jira tasks. We could also have utilized webhooks on Jira itself that would updated targeted issue without an overhead of looping over all the tasks every N minutes.

akbarkz replied:
Definitely that would be a better approach, but that would also be way more complex to implement. I suggest we create a task for the next cycle and address it then. For now I don't think we will have so many Jira tasks to worry about performance of this approach.

akbarkz commented on static/client/components/JiraTasks/JiraTasks.tsx:
```diff
@@ -1,12 +1,19 @@
 
 const JiraTasks = ({ tasks }: IJiraTasksProps): JSX.Element => {
+  const sortedTasks = useMemo(
+    () => tasks.sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime()),
+    [tasks],
+  );
+
   return (
     <table>
-      {tasks.map((task) => (
+      {sortedTasks.map((task) => (
```
That's a good one! :+1:

muhammad-ali-pk reviewed (approved):
LGTM! Approved!
